DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/Transpose.hpp

```diff
@@ -5,65 +5,195 @@
+          typename MatrixOutPipe  // Send the output matrix in column order.
+                                  // Send k_pipe_width elements of type T
+                                  // wrapped in NTuple on each write.
+          >
+event SubmitTransposeKernel(queue& q) {
   // Template parameter checking
   static_assert(std::numeric_limits<short>::max() > k_num_cols_in,
-    "k_num_cols_in must fit in a short" );
-  static_assert( k_num_cols_in % k_pipe_width == 0,
-    "k_num_cols_in must be evenly divisible by k_pipe_width" );
-
-  using PipeType = NTuple< T, k_pipe_width >;
-
-  auto e = q.submit([&](handler& h) {
-
-    h.single_task< TransposeKernelName > ( [=] {
-
-      while( 1 ) {
-
-        // TODO
-        // for now, DO NOT do the transpose, just accept the data already in
-        // column order and pass it through
-        for ( short col = 0; col < k_num_cols_in; col++ ) {
-          PipeType data_in = MatrixInPipe::read();
-          MatrixOutPipe::write( data_in );
+                "k_num_cols_in must fit in a short");
+  static_assert(k_num_cols_in % k_pipe_width == 0,
+                "k_num_cols_in must be evenly divisible by k_pipe_width");
+
+  return q.submit([&](handler& h) {
+    h.single_task<TransposeKernelName>([=]() {
+      // start the transposer
+      Transposer<T, k_num_cols_in, k_pipe_width, MatrixInPipe, MatrixOutPipe>
+          TheTransposer;
+      TheTransposer();
+    });
+  });
+}
+
+// The generic transpose. We use classes here because we want to do partial
+// template specialization on the 'k_pipe_width' to optimize the case where
+// 'k_pipe_width' == 1 and this cannot be done with a function.
+template <typename T, size_t k_num_cols_in, size_t k_pipe_width,
+          typename MatrixInPipe, typename MatrixOutPipe>
+struct Transposer {
+  void operator()() const {
+    using PipeType = NTuple<T, k_pipe_width>;
+
+    // This is a scratch pad memory that we will use to do the transpose.
+    // We read the data in from a pipe (k_pipe_width elements at at time),
+    // store it in this memory in row-major format and read it out in
+    // column-major format (again, k_pipe_width elements at a time).
+    constexpr int kNumScratchMemCopies = 4;
+    constexpr unsigned char kNumScratchMemCopiesBitMask = 0x03;
+    constexpr int kBankwidth = k_pipe_width * sizeof(T);
+    // NO-FORMAT comments are for clang-format
+    [[intel::numbanks(1)]]                   // NO-FORMAT: Attribute
+    [[intel::bankwidth(kBankwidth)]]         // NO-FORMAT: Attribute
+    [[intel::private_copies(1)]]             // NO-FORMAT: Attribute
+    [[intel::max_replicates(k_pipe_width)]]  // NO-FORMAT: Attribute
+    T scratch[kNumScratchMemCopies][k_pipe_width][k_num_cols_in];
+
+    // track the status of each of the buffers
+    // NO-FORMAT comments are for clang-format
+    [[intel::fpga_register]]  // NO-FORMAT: Attribute
+    bool ready_to_send[kNumScratchMemCopies] = {false, false, false, false};
+
```

Comment:
Why do you use registers here? but and `NTuple` for `almost_full_pipe`

Reply:
I use registers for ready_to_send because I want to access it with a (non-static) variable as the index.

I use NTuple for almost_full_pipe because I was worried about creating a false loop carried dependency and having to add an ivdep to the loop.